protocol/clproto_cpp/README.md

```diff
@@ -0,0 +1,47 @@
+# clproto_cpp - Control Libraries Protocol (C++)
+
+This library provides a simple interface for encoding and decoding
+control library type objects to and from a serialized binary string
+representation (the wire format).
+
+```c++
+#include <clproto.h>
+#include <state_representation/space/cartesian/CartesianState.hpp>
+
+auto state = state_representation::CartesianState::Random("frame");
+
+// Encode the message into the wire format
+std::string message = clproto::encode(state);
+
+// The type is automatically derived from the object,
+// but can also be provided explicitly
+clproto::encode<state_representation::CartesianState>(state);
```

Comment:
does this mean you could encore a `CartesianPose` as a `CartesianState` encoded string? like clproto::encode<CartesianState>(CartesianPose::Random("test"))`? I guess the opposite will throw an encoding error because of some of the extra fields right?

Reply:
Yes to the first question and no the second. The opposite case is probably more useful:
`clproto::encode<CartesianPose>(CartesianState::Random("test"))`

This is to allow some polymorphism during encoding. If you are upcasting from Pose to State, then message will be sent as a State but the velocity and twist members will be default initialised, which maybe isn't so useful for the recipient. But by "downcast" using the Pose constructor from State, you can choose to send only the Pose data, and the message will explicitly be a Pose message (not decodable as a State). On the receiving end you can always upcast the decoded Pose back to a State object if you want.

Reply:
Ok this is interesting then. Perfect